lib/CL/devices/tbb/tbb_scheduler.cc

```diff
@@ -0,0 +1,367 @@
+    uint32_t position = 0;
+    pc.printf_buffer_position = &position;
+    assert (pc.printf_buffer != NULL);
+    assert (pc.printf_buffer_capacity > 0);
+    assert (pc.printf_buffer_position != NULL);
+
+    /* Flush to zero is only set once at start of kernel (because FTZ is
+     * a compilation option), but we need to reset rounding mode after every
+     * iteration (since it can be changed during kernel execution). */
+    unsigned flush = k->kernel->program->flush_denorms;
+    if (current_ftz != flush) {
+      pocl_set_ftz (flush);
+      current_ftz = flush;
+    }
+
+    for (size_t x=r.pages().begin(); x!=r.pages().end(); x++) {
+      for (size_t y=r.rows().begin(); y!=r.rows().end(); y++) {
+        for (size_t z=r.cols().begin(); z!=r.cols().end(); z++) {
+          pocl_set_default_rm();
+          k->workgroup((uint8_t *) arguments, (uint8_t * ) & pc, x, y, z);
+        }
+      }
+    }
+
+    if (position > 0) {
+      write (STDOUT_FILENO, pc.printf_buffer, position);
+    }
+
+    free_kernel_arg_array_with_locals ((void **)&arguments, (void **)&arguments2, k);
+  }
+  WorkGroupScheduler( kernel_run_command *k ) :
+      my_k(k)
+  {}
+};
+
+static void
+finalize_kernel_command (kernel_run_command *k)
+{
+  free_kernel_arg_array (k);
+
+  pocl_release_dlhandle_cache (k->cmd);
+
+  pocl_ndrange_node_cleanup (k->cmd);
+
+  POCL_UPDATE_EVENT_COMPLETE_MSG (k->cmd->event, "NDRange Kernel        ");
+
+  pocl_mem_manager_free_command (k->cmd);
+  free_kernel_run_command (k);
+}
+
+static kernel_run_command *
+pocl_tbb_prepare_kernel (void *data, _cl_command_node *cmd)
+{
+  kernel_run_command *run_cmd;
+  cl_kernel kernel = cmd->command.run.kernel;
+  struct pocl_context *pc = &cmd->command.run.pc;
+
+  pocl_check_kernel_dlhandle_cache (cmd, 1, 1);
+
+  run_cmd = new_kernel_run_command ();
+  run_cmd->data = data;
+  run_cmd->kernel = kernel;
+  run_cmd->device = cmd->device;
+  run_cmd->pc = *pc;
+  run_cmd->cmd = cmd;
+  run_cmd->pc.printf_buffer = NULL;
+  run_cmd->pc.printf_buffer_capacity = scheduler.printf_buf_size;
+  run_cmd->pc.printf_buffer_position = NULL;
+  run_cmd->num_groups[0] = pc->num_groups[0];
+  run_cmd->num_groups[1] = pc->num_groups[1];
+  run_cmd->num_groups[2] = pc->num_groups[2];
+  run_cmd->workgroup = reinterpret_cast<pocl_workgroup_func> (cmd->command.run.wg);
+  run_cmd->kernel_args = cmd->command.run.arguments;
+  run_cmd->next = NULL;
+
+  setup_kernel_arg_array (run_cmd);
+
+  pocl_update_event_running (cmd->event);
+
+  return (run_cmd);
+}
+
+static _cl_command_node *
+check_cmd_queue_for_device ()
+{
+  _cl_command_node *cmd;
+  DL_FOREACH (scheduler.work_queue, cmd)
+  {
+    DL_DELETE (scheduler.work_queue, cmd)
+    return cmd; // return first cmd, TODO Is the double linked list still necessary?
+  }
+
+  return NULL;
+}
+
+static void
+tbb_exec_command (kernel_run_command *run_cmd)
+{
+  /* TODO Allow grain size variation for each dimension individually? */
```

Comment:
Is there are scenario where code could benefit? How much complexity would it add? Make a decision